### PR TITLE
feat: add payment composable

### PR DESCRIPTION
**Pull request by wa0x6e**
### Summary

Depends on https://github.com/snapshot-labs/sx-monorepo/pull/1174
Toward https://github.com/snapshot-labs/workflow/issues/420

Continuing on the previous PR, this PR will connect the self-serve page to the payment contract

The process consist of multiple steps, and will explain and guide the user through each step:

- show a modal to choose the payment tokens (usdc/usdt) (will depend on env: testnet and mainnet have different tokens)
- check balance: prevent form submission it the wallet does not have the required balance
- check approval: check if the user has approve the spending on the token contract, else will ask for it
- approve: ask the user to sign the approval spending transaction
- check approval (again): check again for the approval, and go back to approve step if not approved, else continue to next step
- pay: call the pay method on the schnaps contract
- after payment: show success message, your space is now turbo

### How to test

1. Add `VITE_METADATA_NETWORK=s-tn` to your .env.local (to use testnet payment token)
2. Go to a offchain space turbo page (need to append /pro to the url)
3. Click on the "Upgrade" button (on turbo space, is should say "extend"), and the end date in the modal should take into account current subscription it any
4. Follow the modal instruction to start the payment process

### TODO

- [x] Handle errors when wallet is not supported (argent x)
- [ ] form errors are lost after opening the token picker (global to all token picker)
- [x] fetch current turbo plan expiration date

### CODE REVIEW

**Comment by wa0x6e** [comment]
Still a few edge cases to fix, but the main workflow should be ready to test @less 

To test with testnet tokens, set in your .env

```
VITE_METADATA_NETWORK=s-tn
```

File: apps/ui/src/views/Space/Turbo.vue [hunk 1]
```diff
@@ -1,20 +1,36 @@
 <script setup lang="ts">
+import { TOKENS } from '@/composables/usePayment';
 import { _n } from '@/helpers/utils';
+import { metadataNetwork } from '@/networks';
+import { ChainId, Space } from '@/types';
 import ICInfinity from '~icons/c/infinity.svg';
 import ICPro from '~icons/c/pro.svg';
 import ICCheck from '~icons/heroicons-outline/check.vue';
```
Comment (bonustrack) [inline apps/ui/src/views/Space/Turbo.vue:8]: There is an issue here that block dev script to run, should remove .vue. Also actually we dont need to import the icon they are available like this <IS-check />

File: apps/ui/src/composables/usePayment.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+const PAYMENT_CONTRACT_ABI = [
+  'function payWithERC20Token(address token, uint256 amount, bytes barcode)'
+] as const;
+
+export const ASSETS = {
+  USDC: {
+    decimal: 6,
+    symbol: 'USDC',
+    logo: 'ipfs://bafybeienhj4ymzglig6at2bovugjhb6rj4xlfloj2rc4lqad57zwi4zs54'
```
Comment (bonustrack) [inline apps/ui/src/composables/usePayment.ts:32]: We should use Stamp token image resolver here instead.

Comment (wa0x6e) [reply apps/ui/src/composables/usePayment.ts:32]: Stamp does not return any result for base and all sepolia tokens

Comment (bonustrack) [reply apps/ui/src/composables/usePayment.ts:32]: It works, check here, maybe you miss the network prefix https://snapshot.box/#/base:0x282d013BF31374D0046F04D6a7644d97751ed339/treasury/1/tokens

Comment (wa0x6e) [reply apps/ui/src/composables/usePayment.ts]: localhost_8080_

Was missing the network prefix yes, but logo still missing on testnet, but no big deal

**Comment by bonustrack** [comment]
Can you set same pricing on testnet (600/mo 6k/y), and use this token on Sepolia: 
https://sepolia.etherscan.io/address/0x8EC1409bF214893aEA175C9C00c711593277F1C3

**Comment by wa0x6e** [comment]
> Can you set same pricing on testnet (600/mo 6k/y), and use this token on Sepolia: https://sepolia.etherscan.io/address/0x8EC1409bF214893aEA175C9C00c711593277F1C3

done

**Comment by ChaituVR** [comment]
Not important but maybe we should move Monthly/Yearly selection inside payment modal. 

because if i click here, i am not sure if i select montly or other
image

File: apps/ui/src/components/Modal/Payment.vue [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+  (e: 'confirmed');
+}>();
+
+const { auth } = useWeb3();
+const {
+  start: startPaymentProcess,
+  goToNextStep,
+  isLastStep,
+  currentStep
+} = usePaymentFactory(props.network);
```
Comment (Sekhmet) [inline apps/ui/src/components/Modal/Payment.vue:42]: I don't think it matter here, but we are passing non-reactive `network`.

File: apps/ui/src/components/Modal/Payment.vue [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+      : null;
+  })
+});
+
+const selectedTokenAddress = ref<string>('');
+const showPicker = ref(false);
+const hidden = ref(false);
+const modalTransactionProgressOpen = ref(false);
+const isTermsAccepted = ref(false);
+const form = ref(clone(FORM));
```
Comment (Sekhmet) [inline apps/ui/src/components/Modal/Payment.vue:59]: We use `isTermAccepted` so maybe we can use `isHidden` and `isModalTransactionProgressOpen` to match.

File: apps/ui/src/components/SpaceDelegates.vue [hunk 5]
```diff
@@ -460,7 +460,10 @@ watchEffect(() => setTitle(`Delegates - ${props.space.name}`));
         :open="isUndelegating"
         :chain-id="delegation.chainId"
         :execute="undelegateFn"
-        @confirmed="handleUndelegateConfirmed"
+        @confirmed="
+          handleUndelegateConfirmed();
+          isUndelegating = false;
+        "
```
Comment (ChaituVR) [inline apps/ui/src/components/SpaceDelegates.vue:466]: Not important but instead of here, better to move `isUndelegating = false;` inside `handleUndelegateConfirmed`

File: apps/ui/src/components/Modal/Vote.vue [hunk 6]
```diff
@@ -247,7 +247,10 @@ watchEffect(async () => {
         approveTitle: 'Confirm vote'
       }"
       :execute="voteFn"
-      @confirmed="handleConfirmed"
+      @confirmed="
+        handleConfirmed();
+        modalTransactionOpen = false;
+      "
```
Comment (ChaituVR) [inline apps/ui/src/components/Modal/Vote.vue:253]: Same here

File: apps/ui/src/views/Space/Pro.vue [hunk 7]
```diff
@@ -129,37 +210,43 @@ const features = computed<Feature[]>(() => {
...
+        </button>
       </div>
-      <UiButton class="primary">Upgrade {{ space.name }}</UiButton>
+      <UiButton
+        class="primary"
+        :disabled="space.network !== metadataNetwork"
+        @click="handleTurboClick"
+      >
+        {{ space.turbo ? 'Extend' : 'Upgrade' }} {{ space.name }}
+      </UiButton>
```
Comment (ChaituVR) [inline apps/ui/src/views/Space/Pro.vue:249]: Would be nice to show current expiry somewhere here

File: apps/ui/src/views/Space/Pro.vue [hunk 8]
```diff
@@ -203,41 +290,44 @@ const features = computed<Feature[]>(() => {
...
     <div class="text-center shapes py-10 bg-skin-border/40 px-4 space-y-4">
       <h2 class="text-[32px]">Get started today</h2>
-      <UiButton class="primary">Upgrade {{ space.name }}</UiButton>
+      <UiButton
+        class="primary"
+        :disabled="space.network !== metadataNetwork"
+        @click="handleTurboClick"
+      >
+        {{ space.turbo ? 'Extend' : 'Upgrade' }} {{ space.name }}
+      </UiButton>
```
Comment (ChaituVR) [inline apps/ui/src/views/Space/Pro.vue:330]: I already "got started" 😄 
image

File: apps/ui/src/components/Modal/TransactionProgress.vue [hunk 9]
```diff
@@ -131,11 +129,20 @@ watch(
...
+          v-else
           class="font-semibold text-skin-heading text-lg"
           v-text="text.title"
         />
-        <div v-text="text.subtitle" />
+        <slot
+          v-if="$slots.successSubtitle && step === 'success'"
+          name="successSubtitle"
+        />
+        <div v-else v-text="text.subtitle" />
```
Comment (Sekhmet) [inline apps/ui/src/components/Modal/TransactionProgress.vue:145]: I feel like it's getting too convoluted now with props having `successTitle`/`successSubtitle` and now also slots for those as well. It' also incomplete, if we want to override components for `approve`/`confirming`/`fail` steps we will need to keep adding slots.

I think it makes sense for consumer to just take over this entire section if we are already adding so much customization:
```diff
diff --git a/apps/ui/.env b/apps/ui/.env
index 9f8b07c85..cafd1eb6a 100644
--- a/apps/ui/.env
+++ b/apps/ui/.env
@@ -1,5 +1,5 @@
 VITE_ENABLED_NETWORKS=
-VITE_METADATA_NETWORK=
+VITE_METADATA_NETWORK=s-tn
 VITE_MANA_URL=https://mana.box
 VITE_WHITELIST_SERVER_URL=https://wls.snapshot.box
 VITE_HIGHLIGHT_URL=https://testnet.highlight.red
diff --git a/apps/ui/src/components/Modal/Payment.vue b/apps/ui/src/components/Modal/Payment.vue
index 92db91a34..34fb8b3e5 100644
--- a/apps/ui/src/components/Modal/Payment.vue
+++ b/apps/ui/src/components/Modal/Payment.vue
@@ -278,17 +278,24 @@ watch(
       modalTransactionProgressOpen = false;
     "
   >
-    <template #successTitle>
-      <slot name="transactionModalSuccessTitle">
+    <template #headerContent="{ step, text }">
+      <template v-if="step === 'success'">
+        <slot name="transactionModalSuccessTitle">
+          <h4
+            class="font-semibold text-skin-heading text-lg flex flex-col items-center gap-2 mb-3"
+          >
+            Payment successful
+          </h4>
+          <slot name="transactionModalSuccessSubtitle" />
+        </slot>
+      </template>
+      <template v-else>
         <h4
-          class="font-semibold text-skin-heading text-lg flex flex-col items-center gap-2 mb-3"
-        >
-          Payment successful
-        </h4>
-      </slot>
-    </template>
-    <template #successSubtitle>
-      <slot name="transactionModalSuccessSubtitle" />
+          class="font-semibold text-skin-heading text-lg"
+          v-text="text.title"
+        />
+        <div v-text="text.subtitle" />
+      </template>
     </template>
   </ModalTransactionProgress>
 </template>
diff --git a/apps/ui/src/components/Modal/TransactionProgress.vue b/apps/ui/src/components/Modal/TransactionProgress.vue
index ec242c66b..3b085c953 100644
--- a/apps/ui/src/components/Modal/TransactionProgress.vue
+++ b/apps/ui/src/components/Modal/TransactionProgress.vue
@@ -129,20 +129,13 @@ watch(
         <IS-x-mark :width="28" :height="28" />
       </div>
       <div class="flex flex-col space-y-1 leading-6">
-        <slot
-          v-if="$slots.successTitle && step === 'success'"
-          name="successTitle"
-        />
-        <h4
-          v-else
-          class="font-semibold text-skin-heading text-lg"
-          v-text="text.title"
-        />
-        <slot
-          v-if="$slots.successSubtitle && step === 'success'"
-          name="successSubtitle"
-        />
-        <div v-else v-text="text.subtitle" />
+        <slot name="headerContent" :step="step" :text="text">
+          <h4
+            class="font-semibold text-skin-heading text-lg"
+            v-text="text.title"
+          />
+          <div v-text="text.subtitle" />
+        </slot>
       </div>
     </div>
     <slot id="content" :step="step" :tx-id="txId" />
```

What do you think?

File: apps/ui/src/composables/usePayment.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+function getWeiAmount(token: Token, amount: number): BigNumber {
+  return BigNumber.from(amount * 10 ** token.decimals);
+}
+
+export default function usePayment(network: ChainId) {
+  const { auth } = useWeb3();
+  const { modalAccountOpen } = useModal();
+
+  async function hasApproved(
```
Comment (Sekhmet) [inline apps/ui/src/composables/usePayment.ts:62]:
```suggestion
  async function getIsApproved(
```

I think we discussed using `get` prefix so it's clear it's function.

**Review by Sekhmet — COMMENTED** [review]
Other than that it works good to me. I tested the flow.

File: apps/ui/src/views/Space/Pro.vue (hunk 7 above)
Comment (wa0x6e) [reply apps/ui/src/views/Space/Pro.vue:249]: @bonustrack

File: apps/ui/src/components/AppLink.vue [hunk 11]
```diff
@@ -35,10 +36,22 @@ function normalize(to: RouteLocationRaw) {
...
   return to;
 }
+
+function normalizeExternalize(to: RouteLocationRaw | string): string {
+  if (typeof to === 'string') {
+    return to;
+  }
+
+  return new URL(router.resolve(to).href, window.location.origin).href;
+}
```
[the remark this answers is not in the document]
Comment (wa0x6e) [reply apps/ui/src/components/AppLink.vue:46]: This method either take a string (already absolute url), or a route, and convert it to an absolute url.

File: apps/ui/src/views/Space/Pro.vue (hunk 7 above)
Comment (bonustrack) [reply apps/ui/src/views/Space/Pro.vue:249]: We will make it into another page later

File: apps/ui/src/components/AppLink.vue (hunk 11 above)
[the remark this answers is not in the document]
Comment (Sekhmet) [reply apps/ui/src/components/AppLink.vue:46]:
```diff
diff --git a/apps/ui/src/components/AppLink.vue b/apps/ui/src/components/AppLink.vue
index f3be6510b..317fee844 100644
--- a/apps/ui/src/components/AppLink.vue
+++ b/apps/ui/src/components/AppLink.vue
@@ -42,7 +42,7 @@ function normalizeExternalize(to: RouteLocationRaw | string): string {
     return to;
   }
 
-  return new URL(router.resolve(to).href, window.location.origin).href;
+  return router.resolve(to).href;
 }
 </script>
 
```

Works as well, unless we really want to have `window.location.origin` in the URL, but for our usage it's not needed.

[the remark this answers is not in the document]
Comment (Sekhmet) [reply apps/ui/src/components/AppLink.vue:46]: I will also call it `resolveLocation` or `resolveToUrl`.

**Review by ChaituVR — APPROVED** [review]
tAck
